### PR TITLE
Frontend refactor: typescript added, updated packages,  directory structure and functional components

### PR DESCRIPTION
**Pull request by lakshz**
typescript added, updated packages, improved directory structure, and converted all class-based components to functional components

Completes task 1, 2 and 3 of  issue #28

### CODE REVIEW

File: frontend/src/components/CustomTable/CustomTable.js
Comment (chirag-ghosh) [inline frontend/src/components/CustomTable/CustomTable.js:1]: This and `frontend/src/components/CustomTable/CustomTableRow.js` can be in the same file and exported as different named components.

File: frontend/src/components/EmptyRoomTable/EmptyRoomsTable.js
Comment (chirag-ghosh) [inline frontend/src/components/EmptyRoomTable/EmptyRoomsTable.js:1]: Same with this and `frontend/src/components/EmptyRoomTable/EmptyRoomsTableRow.js`

File: frontend/src/components/TwoSlotDisplay.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
+import React, { useReducer } from "react";
+import PropTypes from "prop-types";
+import EmptyRooms from "./EmptyRooms.js";
+import { Slots, DayNames, Complexes, Floors } from "../constants/constants.js";
+import { getNextSlot, getInitialChillPlaceDetails } from "../util/utilities.js";
+import { CHILLPLACE_DISPATCH_TYPES } from "../constants/constants.js";
+
+const chillPlaceDetailsReducer = (state, action) => {
```
Comment (chirag-ghosh) [inline frontend/src/components/TwoSlotDisplay.js:8]: What is the need to use `chillPlace` in the function name? If needed we can use `chillzone` (name of repository). here even that is not necessary. Even constant type names do not need it.

File: frontend/src/constants/constants.js [hunk 2]
```diff
@@ -47,3 +47,10 @@ export const Floors = {
   3: "3",
   4: "4",
 };
+
+export const CHILLPLACE_DISPATCH_TYPES = {
```
Comment (chirag-ghosh) [inline frontend/src/constants/constants.js:51]: can just use `DISPATCH_TYPES`

**Review by chirag-ghosh — CHANGES_REQUESTED** [review]

File: frontend/src/components/CustomTable/CustomTable.js
Comment (LakshyaSatpal) [reply frontend/src/components/CustomTable/CustomTable.js:1]: Done.

File: frontend/src/constants/constants.js (hunk 2 above)
Comment (LakshyaSatpal) [reply frontend/src/constants/constants.js]: Done.

File: frontend/src/components/TwoSlotDisplay.js (hunk 1 above)
Comment (LakshyaSatpal) [reply frontend/src/components/TwoSlotDisplay.js]: Okay, I named it as just `reducer` now. Does this work?

File: frontend/src/index.tsx [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import App from "./App";
+import registerServiceWorker from "./registerServiceWorker";
+import schedule from "./data/schedule.json";
+import emptySchedule from "./data/empty_schedule.json";
+
+const rootElement = document.getElementById("root")!;
+rootElement.className = "container-fluid";
+
+const root = createRoot(rootElement);
+root.render(<App schedule={schedule} emptySchedule={emptySchedule} />);
```
Comment (chirag-ghosh) [inline frontend/src/index.tsx:13]: import and use `schedule` and `emptySchedule` json files directly in App.tsx. I see no point in passing them as props here.

File: frontend/src/registerServiceWorker.ts [hunk 4]
```diff
@@ -9,28 +9,28 @@
...
     window.location.hostname.match(
       /^127(?:\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}$/
     )
 );
 
 export default function register() {
-  if (process.env.NODE_ENV === 'production' && 'serviceWorker' in navigator) {
+  if (process.env.NODE_ENV === "production" && "serviceWorker" in navigator) {
     // The URL constructor is available in all browsers that support SW.
-    const publicUrl = new URL(process.env.PUBLIC_URL, window.location);
```
Comment (chirag-ghosh) [inline frontend/src/registerServiceWorker.ts:24]: why is `window.location` removed?

File: frontend/src/util/utilities.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    }
+  }
+
+  // if (slot === undefined) {
+  //   let hour = today.getHours();
+  //   slot = 0;
+  //   if (hour >= 18) {
+  //     day = (day + 1) % 9;
+  //   }
+  // }
```
Comment (chirag-ghosh) [inline frontend/src/util/utilities.ts:53]: remove if not necessary now

File: frontend/src/util/utilities.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    prev_day = (day - 1) % 5;
+  }
+
+  return {
+    day: prev_day,
+    slot: (slot - 1) % 9,
+  };
+}
+
+export function getInitialChillPlaceDetails(): ChillPlaceDetails {
```
Comment (chirag-ghosh) [inline frontend/src/util/utilities.ts:28]: use function name `getInitialChillzoneDetails`

File: frontend/src/lib/types.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+import schedule from "../data/schedule.json";
+import emptySchedule from "../data/empty_schedule.json";
+
+export type Schedule = typeof schedule;
+
+export type SubjectDetails = typeof subjectDetails;
+
+export type EmptySchedule = typeof emptySchedule;
+
+export type Complex = "Any" | "NC" | "NR" | "Vikramshila";
```
Comment (chirag-ghosh) [inline frontend/src/lib/types.ts:11]: Why do we need `"Any"` type?

File: frontend/src/lib/types.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+export type Floor = "Any" | "1" | "2" | "3" | "4";
+
+export type ChillPlace = {
```
Comment (chirag-ghosh) [inline frontend/src/lib/types.ts:14]: Can use a better type name like `Classroom`

P.S., I don't see this type used anywhere. Let's merge `isWeekend` attribute here itself and rename to `Classroom`

File: frontend/src/lib/types.ts [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+  day: number;
+  slot: number;
+  floor: Floor;
+  complex: Complex;
+};
+
+export type ChillPlaceDetails = ChillPlace & {
```
Comment (chirag-ghosh) [inline frontend/src/lib/types.ts:21]: Can remove this.

File: frontend/src/lib/types.ts (hunk 7 above)
Comment (LakshyaSatpal) [reply frontend/src/lib/types.ts:11]: "Any" is the intial value that we use for complex.

File: frontend/src/registerServiceWorker.ts (hunk 4 above)
Comment (LakshyaSatpal) [reply frontend/src/registerServiceWorker.ts:24]: URL constructor takes second argument as base url for the case if first arg is relative url. But, `process.env.PUBLIC_URL` in itself is an absolute url, so no need for second arg.

**Review by chirag-ghosh — APPROVED** [review]
LGTM

**Comment by proffapt** [comment]
@chirag-ghosh what about this?

**Comment by chirag-ghosh** [comment]
Have to check if the frontend works. iirc Shikhar had some problems making
it work

**Comment by proffapt** [comment]
@shikharish ^^

**Comment by shikharish** [comment]
Yes, I ran it locally, there was this issue
image

**Comment by chirag-ghosh** [comment]
Fixed it now. Although need to fix the conflicts too

**Review by shikharish — APPROVED** [review]
LGTM

**Pull request merged**